lightning/src/onion_message/messenger.rs

```diff
@@ -1320,42 +1345,57 @@ where
 	pub async fn process_pending_events_async<Future: core::future::Future<Output = Result<(), ReplayEvent>> + core::marker::Unpin, H: Fn(Event) -> Future>(
 		&self, handler: H
 	) {
-		let mut intercepted_msgs = Vec::new();
-		let mut peer_connecteds = Vec::new();
-		{
-			let mut pending_intercepted_msgs_events =
-				self.pending_intercepted_msgs_events.lock().unwrap();
-			let mut pending_peer_connected_events =
-				self.pending_peer_connected_events.lock().unwrap();
-			core::mem::swap(&mut *pending_intercepted_msgs_events, &mut intercepted_msgs);
-			core::mem::swap(&mut *pending_peer_connected_events, &mut peer_connecteds);
+		if self.pending_events_processor.compare_exchange(false, true, Ordering::Acquire, Ordering::Relaxed).is_err() {
+			return;
 		}
 
-		let mut futures = Vec::with_capacity(intercepted_msgs.len());
-		for (node_id, recipient) in self.message_recipients.lock().unwrap().iter_mut() {
-			if let OnionMessageRecipient::PendingConnection(_, addresses, _) = recipient {
-				if let Some(addresses) = addresses.take() {
-					futures.push(handler(Event::ConnectionNeeded { node_id: *node_id, addresses }));
+		{
+			let intercepted_msgs = self.pending_intercepted_msgs_events.lock().unwrap().clone();
+			let mut futures = Vec::with_capacity(intercepted_msgs.len());
+			for (node_id, recipient) in self.message_recipients.lock().unwrap().iter_mut() {
+				if let OnionMessageRecipient::PendingConnection(_, addresses, _) = recipient {
+					if let Some(addresses) = addresses.take() {
+						futures.push(handler(Event::ConnectionNeeded { node_id: *node_id, addresses }));
+					}
 				}
 			}
-		}
 
-		for ev in intercepted_msgs {
-			if let Event::OnionMessageIntercepted { .. } = ev {} else { debug_assert!(false); }
-			futures.push(handler(ev));
+			// The offset in the `futures` vec at which `intercepted_msgs` start. We don't bother
+			// replaying `ConnectionNeeded` events.
+			let intercepted_msgs_offset = futures.len();
```

Comment:
Would it be simpler to use a separate `MultiEventFuturePoller`?

Reply:
Hmmm... right. Would using some hierarchical approach work? Less code would be better, IMO.

Reply:
A poller that contains two pollers -- one for each event type.

Reply:
Right, might be more trouble than it is worth. I was thinking after the poller completed you would break it back into its two parts to inspect the results separately. But I suppose the overall poller would still need *some* result. Maybe the overall poller could just be a `Future` implementation on a tuple of `Future`s / pollers? Anyhow, feel free to punt on this if it's more complicated.